### ISSUE TITLE
list.count function returns incorrect value when list is empty

### ISSUE THREAD

**Issue by alexspurling** [onyx-lang/onyx#83]
The following loops forever on (c86358ca) and gives an "index out of bounds" error in the current Playground version:

```
use core.list
use core {println, printf}

main :: () {
    l := list.make(i32);
    
    list.push_begin(&l, 5);
    list.push_begin(&l, 3);
    list.push_begin(&l, 1);
    
    println(l);
    
    println(list.count(&l));
    
    while list.count(&l) > 0 {
        last := list.pop_end(&l);
        printf("Last item: {}\n", last);
        println(l);
        println(list.count(&l));
    }
}
```

Inspecting the list after the last `pop_end` call, you can see that the `last` pointer is null, but the `first` pointer is unchanged. The ListElem struct that `first` points to also points to itself so the count function just loops forever.

If we're fixing bugs in `list`, it might be worth adding an `empty` function at the same time.

**Comment by brendanfh** [comment]
Re-reading the `core.list` implementation is almost embarrassing. I made more mistakes in it than I did in my college data structures course. I will rewrite most of the procedures in `core.list` to close this issue.

**Comment by brendanfh** [comment]
I have fixed this in [32c8f94](https://github.com/onyx-lang/onyx/commit/32c8f94a8584ffc02b3bfe56c7aca1a86d888756), with changes inspired from your fork. Thanks for submitting the issue!
